### DISCUSSION TITLE
Best practice for simple POST handling

### DISCUSSION THREAD

**Discussion by csbence10** [rrousselGit/riverpod#2790] [Q&A] [answered]
Hello @rrousselGit,

I've been wondering about this for a long time. How do you guys handle simple posting requests (like sending a form, commenting on a post, sending a like) using riverpod, while keeping your code simple and also updating the UI and having side effects.
The imaginary requirements are:
- simple form with `name` and `description` params, call it `FormState`
- the form is sent with a press of a button
- validate the form (let's say the name prop is required only)
- handle loading state (like disabling the button and having an activity indicator at the center, instead of a "Send" text)
- handle error, show a snack bar/dialog if the API call failed
- on success pop the form screen

I have 3 approaches in mind:

### Sync notifier

1. have a callback for each textfield's `onChange`, and update the properties 
2. extend the state with `isLoading`, `error, `success` properties (or use sealed class, or freezed)
3. we can validate the form having a computed prop depending on `name`
4. on button tap we emit a new state, setting `isLoading` to `true`, call the API, use a `try catch` block to handle error. Finally we set our success/error fields depending on the result of the call and setting the loading state back
5. on the UI now we can `watch` the validation, and the loading state
6. also we can listen to the changes and show a dialog for errors and pop our route if we succeeded.

**Note:** Instead of having a `success` flag we could make our `sendForm` method the return a bool indicating the result of the API call, so we could have handle the navigation in the `onTap` handler. I'm not a fan of this. As the UI is the function of the state, this side effects should depend on the state as well.

This approach seems quite straightforward but we definitely have boilerplate code, and always settings the flags is error prone (especially if we have something more complex)

### Async notifier

1. same as before, but now we should emit something like `state = AsyncData(state.copyWith(...));` which is "unnecessary"
2. Now loading and error states are handled for us by `AsyncLoading` and `AsyncError`, but what about success?
3. we can validate the form having a computed prop depending on `name`
4. On a button tap we emit `AsyncLoading`, again call our API using `try catch` block, or `AsyncValue.guard`. We have our result set the state depending on that (if we used guard we don't even have to set the error state) but how do we know about success? Do I need a property inside `FormState` here as well?
5. The UI is a bit different, we use `when`, `maybeWhen` etc, but it's okay
6. Now listening is another question since we know about error, so the dialog can be shown, but again, what about success, popping the route?

### Async notifier with a router provider

Most of the times I use `package:auto_route` for navigation. I can create a `routerProvider` and read this one in other providers, so this approach is kinda same as the previous one but in point 4. instead after the API call:
- if it was successful I read my router and pop the current route
- if it failed I read my router, get it's context and show a dialog using that context
Therefore I spare some lines of code while setting up a listener in a widget's build method.

I don't really like this last one, it came from the times when I did native iOS development, we always navigated from the view models, but yet this is shortest way to handle these states.

This is such a common task, every app have at least 20 of them, how do you guys approach this problem? Is there even a best practice?

**Comment by rrousselGit** [comment]
See https://github.com/rrousselGit/riverpod/issues/2625 and the upcoming page for it:  https://deploy-preview-2740--river-pod.netlify.app/docs/essentials/side_effects

**Reply by csbence10** [reply]
Nice, thanks for your answer. The solution with hooks looks nice, however I don't use that package. The other one (using a `FutureBuilder` feels a bit mixed)
Do you think having a `FutureProvider` to handle the state changes makes sense?

```dart
final resultProvider = FutureProvider.autoDispose.family((ref, input) {
  return input;
});

@riverpod
class MyNotifier extends _$MyNotifier {
  @override
  Future<List<Todo>> build() {
    return Future.value([]);
  }

  final _values = [
    'Buy milk',
    'Buy eggs',
    'Buy bread',
  ];

  final _random = Random();

  Future<String> addTodo() async {
    await Future.delayed(const Duration(seconds: 1));
    final previousState = await future;
    final randomIndex = _random.nextInt(_values.length);
    final nextItem = _values[randomIndex];
    state = AsyncData([...previousState, Todo(description: nextItem)]);

    return Future.value(nextItem);
  }
}
```

And then the UI:
```dart
class TodoScreen extends ConsumerStatefulWidget {
  const TodoScreen({Key? key}) : super(key: key);

  @override
  ConsumerState<TodoScreen> createState() => _TodoScreenState();
}

class _TodoScreenState extends ConsumerState<TodoScreen> {
  Future<String>? _addTodoFuture;

  void _onTap() {
    setState(() {
      _addTodoFuture = ref.read(myNotifierProvider.notifier).addTodo();
    });
  }

  @override
  Widget build(BuildContext context) {
    ref.listen(resultProvider(_addTodoFuture), (_, state) {
      state.whenData((value) {
        ScaffoldMessenger.of(context)
          ..removeCurrentSnackBar()
          ..showSnackBar(SnackBar(content: Text('$value added')));
      });
    });

    const loader = Center(child: CircularProgressIndicator.adaptive());

    return Scaffold(
      floatingActionButton: ref.watch(resultProvider(_addTodoFuture)).when(
            data: (_) => FloatingActionButton(
              onPressed: _onTap,
              child: const Icon(Icons.add),
            ),
            error: (_, __) => FloatingActionButton(
              onPressed: _onTap,
              child: const Icon(Icons.error),
            ),
            loading: () => FloatingActionButton(
              onPressed: () {},
              child: loader,
            ),
          ),
      body: SafeArea(
        child: ref.watch(myNotifierProvider).when(
              data: (todos) => ListView.builder(
                itemCount: todos.length,
                itemBuilder: (_, index) => ListTile(
                  title: Text(todos[index].description),
                ),
              ),
              error: (error, __) => Text(error.toString()),
              loading: () => loader,
            ),
      ),
    );
  }
}
```

**Reply by rrousselGit** [reply, marked as answer]
That's kind of hacky, but alright I guess.  
I'd still suggest using hooks instead. But I can't think of an immediate issue with such a provider

Although you could use codegen for it too, and should type it!

Riverpod v3 would allow:

```dart
@riverpod
Future<T?> result<T>(ResultRef ref, Future<T>? future>) async => result;
```

With v2, you can remove the generics.
